New option to set a shutdown grace period

bcarlsson opened the pull request:
## What this PR does / why we need it:
When running the ingress controller behind a load balancer in AWS, we need to have a grace period for the shutdown process in case of a scale down event. If not, the pod might be terminated before it has been removed from the load balancer target group.

This PR gives the user an option to set a grace period for when the shutdown phase has been triggered and the actual shutdown begins.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Which issue/s this PR fixes
fixes #4726 

## How Has This Been Tested?
It has been deployed in a running Kubernetes cluster and verified by monitoring the `/healthz` endpoint.

## Checklist:
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/master/CONTRIBUTING.md) guide
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

aledbf commented:
@bcarlsson please add e2e tests.
How this change interacts with the wait command https://github.com/kubernetes/ingress-nginx/blob/master/deploy/static/provider/aws/deploy.yaml#L345 ?

aledbf commented:
/ok-to-test

aledbf commented:
/hold

bcarlsson commented:
> @bcarlsson please add e2e tests.
> How this change interacts with the [wait-shutdown preStop](https://github.com/kubernetes/ingress-nginx/blob/master/deploy/static/provider/aws/deploy.yaml#L345) ?
/wait-shutdown preStop will trigger `/healthz` endpoint to return not ok while `/livez` remains ok during the shutdown grace period.
We'll look into e2e tests.

bcarlsson commented:
@aledbf I've now added e2e tests to verify both endpoints during the shutdown grace period.

bcarlsson commented:
/assign @bowei

aledbf commented:
@bcarlsson please don't assign PRs manually

bcarlsson commented:
@aledbf ok sorry, tried to follow an earlier comment by the robot.

aledbf commented on test/e2e/health/health_endpoints.go:
```diff
@@ -0,0 +1,97 @@
+					args = append(args, v)
+				}
+
+				args = append(args, "--shutdown-grace-period=60")
```
Please use a different value, like 90 or 100
(default terminationGracePeriodSeconds is 30s, configured in ingress-nginx is 300s)

aledbf commented:
@bcarlsson please add the new path in the e2e chart configuration https://github.com/kubernetes/ingress-nginx/blob/master/test/e2e/wait-for-nginx.sh#L71 to make sure we test the feature.

bcarlsson commented:
Thanks for your feedback! 
After investigation, we're no longer sure that the new `/livez` endpoint provides any real value since the pod will already be in a shutting down state and will still be forcefully terminated after TerminationGracePeriodSeconds has elapsed.
The ability to set a shutdown grace period is crucial though, since we need to keep the ingress controller running during the time it will take the AWS load balancer to mark the target as down.

Do you see any advantages of the new `/livez` endpoint that returns status code 200 during the shutdown grace period, or should we remove that part?

jacknagel commented:
I'm confused about the utility of the shutdown grace period. Even if you delay the shutdown of the nginx processes, the endpoints for the terminating pod(s) are going to be removed from the service much sooner than that, right?

maglun2 commented:
@jacknagel I work with @bcarlsson and our use case is that we run a AWS Network Load Balancer that is pointed to nginx ingress controller that is using hostPort.

renan commented on internal/ingress/controller/nginx.go:
```diff
@@ -369,6 +384,7 @@ func (n *NGINXController) Start() {
 // Stop gracefully stops the NGINX master process.
 func (n *NGINXController) Stop() error {
 	n.isShuttingDown = true
+	time.Sleep(time.Duration(n.cfg.ShutdownGracePeriod) * time.Second)
```
Before sleeping we might want to check if the shutdown is already happening. See about 7 lines below:

```
if n.syncQueue.IsShuttingDown() {
...
```

bcarlsson replied:
Good point, I've moved the sleep below IsShuttingDown check. Thank you.

bcarlsson commented:
@aledbf as I wrote in an earlier comment, we do no longer see the point of a new `/livez` endpoint since the pod would already be in a shutting down state.
I've reverted the `/livez` commits and added a test for only the shutdown grace period.

kolorful commented on cmd/nginx/flags.go:
```diff
@@ -171,6 +171,8 @@ Takes the form "<host>:port". If not provided, no admission controller is starte
 
 		statusUpdateInterval = flags.Int("status-update-interval", status.UpdateInterval, "Time interval in seconds in which the status should check if an update is required. Default is 60 seconds")
+
+		shutdownGracePeriod = flags.Int("shutdown-grace-period", 0, "Seconds to wait after receiving the shutdown signal, before stopping the nginx process.")
```
Does the helm templates need to be updated as well so we can customize this value?

aledbf replied:
@kolorful no need, you can use [extraArgs](https://github.com/kubernetes/ingress-nginx/blob/master/charts/ingress-nginx/values.yaml#L120) to set custom flags or values

kolorful replied:
ah nice.

aledbf commented:
@bcarlsson please squash the commits

bcarlsson commented:
@aledbf I've squashed the commits.

jeroenj commented:
It looks like this is good to go? (except for the PR title)

We've been eagerly waiting to get this merged in as on clusters that scale heavily we get some failed requests when the ingress-nginx pods are shutting down but the (GKE) load balancers are still sending traffic to it.

Zhann commented:
What's the state of this? Anything blocking?

aledbf commented:
This change is missing the update of the helm chart values to show how to configure the endpoint and also an e2e test using it.

bcarlsson commented:
@aledbf I apologize for any confusion. As I wrote in a comment earlier on September 14th, we realized that a new `/livez` endpoint did not add any value since the pod would already be in a "shutting down" state and would be terminated after TerminationGracePeriodSeconds had elapsed.

As commented on Oct 7, I did revert all changes for a new `/livez` endpoint but kept the feature to set a grace period before the actual shutdown would begin.

Is a update of the helm chart needed for this as there are no changes to a `/livez` endpoint anymore?

I know that only the last part of the PR description reflects the actual change now. Do you want me to update the title and description?

jeroenj commented:
> I know that only the last part of the PR description reflects the actual change now. Do you want me to update the title and description?

Personally I'd update the description (and title) to reflect the actual change. That said I do think it's gook to keep the initial description around for _historic_ reference as to how this started out and how it came to the current state. That's something we do internally too.

aledbf commented:
> Do you want me to update the title and description?

Yes, please.

aledbf commented:
/test pull-ingress-nginx-e2e-1-19
/test pull-ingress-nginx-e2e-1-20

aledbf commented:
/lgtm
/approve

aledbf commented:
@bcarlsson thanks!

aledbf commented:
@bcarlsson please remove the "Old description for reference" section. There is no code related to that description. Even if that was the initial intention, is confusing.

bcarlsson commented:
> @bcarlsson please remove the "Old description for reference" section. There is no code related to that description. Even if that was the initial intention, is confusing.

Done

aledbf commented:
/hold cancel
